keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1330 @@
+eviction API to be extensible in this way. The webhook approach is therefore not recommended.
+
+As seen in the experience reports and GitHub issues, some admins solve their problems by simply
+ignoring PDBs which can cause unnecessary disruptions or data loss. Some solve this by playing
+with the application deployment, but have to understand that the application supports this.
+
+### Goals
+- Kubectl drain should not evict and disrupt applications with evacuation capability and instead
+  politely ask them to migrate their pods to another node or to remove them by creation of
+  NodeMaintenance object.
+- Introduce a node maintenance controller that will help controllers like deployment controller
+  to migrate their pods.
+- Deployment controller should use `.spec.strategy.rollingUpdate.maxSurge` to evacuate its pods
+  from a node that is under maintenance.
+
+### Non-Goals
+- The PDB controller should detect and account for applications with evacuation capability when
+  calculating PDB status.
+- Introduce a field that could include non-critical daemon set pods
+  (priority `system-cluster-critical` or `system-node-critical`) for node maintenance/drain request.
+  The daemon set controller would then gracefully shut down these pods. Critical pods could be
+  overridden by the priority list mentioned below.
+- NodeMaintenance could include a plan of which pods to target first. Similar to graceful node
+  shutdown, we could include a list of priorities to decide which pods should be terminated first.
+  This list could optionally include pod timeouts, but could also wait for all the pods of a given
+  priority class to finish first without a timeout. This could also be used to target daemon set
+  pods of certain priorities (see point above). We could also introduce drain profiles based on
+  these lists. The cluster admin could then choose or create such a profile based on his/her needs.
+  The logic for processing the decision list would be contained in the node maintenance controller,
+  which would set an intent to selected pods to shut down via the EvacuationRequested
+  condition.
+- Introduce a node maintenance period, nodeDrainTimeout (similar to [cluster-api](https://cluster-api.sigs.k8s.io/developer/architecture/controllers/control-plane)
+  nodeDrainTimeout) or a TTL optional field as an upper bound on the duration of node maintenance.
+  Then the node maintenance would be garbage collected and the node made schedulable again.
+
+## Proposal
+Most of these issues stem from missing a standardized way of detecting a start of the node drain.
+This KEP proposes the introduction of a NodeMaintenance object that would signal an intent to
+gracefully remove pods from given nodes. The application pods should then signal back that the pods
+are being removed or migrated from the node. The implementation should also utilize existing node's
+`.spec.unschedulable` field, which prevents new pods from being scheduled on such a node.
```

Comment:
https://github.com/kubernetes/enhancements/pull/4213/files#r1537798784

This field, can also be manipulated by users... having a dedicated taint with NoSchedule effect would remove that "concurrency/ownership" problem and open for even more control.

Reply:
Let's close that thread since the same discussion is happening in the other comment.